partitionccl: make CreatePartitioning idempotent for implicit columns

otan opened the pull request:
Previously, calling CreatePartitioning twice on an implicitly
partitioned index would un-implicitly partition that index. This commit
changes CreatePartitioning to not un-implicitly partition the index in
such a case.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60696)
<!-- Reviewable:end -->

arulajmani reviewed (changes requested):
Thanks for making this change! 

Couple notes:
- Can we add some tests in partitioning_implicit that test this stuff directly as well     
- Can we cleanup some of the complexity in `alter_primary_key` around altering to/from REGIONAL BY ROW tables? I'm thinking around here: https://github.com/cockroachdb/cockroach/blob/5744288882d60fc9f8e0010edd5c27a4196b158d/pkg/sql/alter_primary_key.go#L450

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60696#-:-MTlWNh69b1RjmZn90k7:b-egs6wa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/ccl/partitionccl/partition.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60696#-MTlXgVe0v555-HLqfFN:-MTlXgVe0v555-HLqfFO:bldaxf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e172b7b41a8af6052d1afbc755bdff603bfd303/pkg/ccl/partitionccl/partition.go#L394)):*
> ```Go
> 			)
> 		}
> 		for i, oldColID := range oldImplicitColumnIDs {
> ```

Just confirming, we don't need to check for the direction of implicit columns right? There's now way that could change?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> Can we add some tests in partitioning_implicit that test this stuff directly as well

What tests would do this? We block everywhere you try to change an implicitly partitioned column when using ALTER INDEX, and that's already tested.

> Can we cleanup some of the complexity in alter_primary_key around altering to/from REGIONAL BY ROW tables? I'm thinking around here:

No, because in that case we *are* changing the implicitly partitioned columns to something else. But we are rewriting the index, so it's okay.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60696#-:-MTlaKHP3oUCLFZZP-jV:b-klvpse)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[pkg/ccl/partitionccl/partition.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60696#-MTlXgVe0v555-HLqfFN:-MTlbQ2w-9Ips85D_tbp:b58dl35) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e172b7b41a8af6052d1afbc755bdff603bfd303/pkg/ccl/partitionccl/partition.go#L394)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Just confirming, we don't need to check for the direction of implicit columns right? There's now way that could change?
</blockquote></details>

Nope -- all implicitly partitioned column is ascending


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
>  We block everywhere you try to change an implicitly partitioned column when using ALTER INDEX
 
I see, I guess I was under the impression that this would no-op and we could test that. I don't think we need to do that in this patch, I'm fine with this as is. 

> No, because in that case we are changing the implicitly partitioned columns to something else. But we are rewriting the index, so it's okay.
I don't know this code quite well, but the comment says 
```we already have the correct implicit
		// column descriptor in front of each index, and calling CreatePartitioning again
		// will make these indexes explicit.
```
The "CreatePartitioning again will make indexes explicit" no longer applies, right? Can't we clean this up? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60696#-:-MTm6WM98LsvgLbszR-3:bhy67mx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)



<!-- Sent from Reviewable.io -->

otan commented:
> The "CreatePartitioning again will make indexes explicit" no longer applies, right? Can't we clean this up?

I see, yes, but that is now needed regardless for the REGIONAL BY ROW -> REGIONAL BY ROW case.

otan committed:
partitionccl: make CreatePartitioning idempotent for implicit columns

Previously, calling CreatePartitioning twice on an implicitly
partitioned index would un-implicitly partition that index. This commit
changes CreatePartitioning to not un-implicitly partition the index in
such a case.

Release note: None

otan commented:
bors r=arulajmani
